### PR TITLE
[system-health] No longer check critical process/service status via monit

### PR DESCRIPTION
**Pull request by Junchao-Mellanox**
HLD updated here: https://github.com/Azure/SONiC/pull/887

#### Why I did it

Command `monit summary -B` can no longer display the status for each critical process, system-health should not depend on it and need find a way to monitor the status of critical processes. The PR is to address that. monit is still used by system-health to do file system check as well as customize check.

#### How I did it

1.	Get container names from FEATURE table
2.	For each container, collect critical process names from file critical_processes
3.	Use “docker exec -it <container_name> bash -c ‘supervisorctl status’” to get processes status inside container, parse the output and check if any critical processes exit

#### How to verify it

1. Add unit test case to cover it
2. Adjust sonic-mgmt cases to cover it
3. Manual test

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [x] 202012
- [x] 202106

### CODE REVIEW

**Comment by Junchao-Mellanox** [comment]
Hi @qiluo-msft , I have fixed all review comment, could you please review and sign-off?

**Comment by Junchao-Mellanox** [comment]
Hi @qiluo-msft , could you please review and sign off?

File: src/system-health/health_checker/service_checker.py [hunk 1]
```diff
@@ -24,23 +43,171 @@ class ServiceChecker(HealthChecker):
 
     def __init__(self):
         HealthChecker.__init__(self)
+        self.critical_process_dict = {}
```
Comment (yozhao101) [inline src/system-health/health_checker/service_checker.py:46]: Do we consider renaming this variable to `container_critical_processes`?

File: src/system-health/health_checker/service_checker.py [hunk 2]
```diff
@@ -24,23 +43,171 @@ class ServiceChecker(HealthChecker):
...
+        self.load_critical_process_cache()
+
+    def get_expected_runnning_container_set(self, feature_table):
+        """Get a set of containers that are expected to running on SONiC
+
+        Args:
+            feature_table (object): FEATURE table in CONFIG_DB
+
+        Returns:
+            set: A set of container names
```
Comment (yozhao101) [inline src/system-health/health_checker/service_checker.py:63]: This function also return a dictionary, right?

File: src/system-health/health_checker/service_checker.py [hunk 3]
```diff
@@ -24,23 +43,171 @@ class ServiceChecker(HealthChecker):
...
+        """
+        containers = set()
```
Comment (yozhao101) [inline src/system-health/health_checker/service_checker.py:65]: Do we consider renaming this variable containers to expected_running_containers?

File: src/system-health/health_checker/service_checker.py [hunk 4]
```diff
@@ -24,23 +43,171 @@ class ServiceChecker(HealthChecker):
...
+        # Containers that has invalid critical_processes file
+        self.bad_containers = set()
+
+        self.container_feature_dict = {}
+
+        self.need_save_cache = False
+
+        self.load_critical_process_cache()
+
+    def get_expected_runnning_container_set(self, feature_table):
```
Comment (yozhao101) [inline src/system-health/health_checker/service_checker.py:56]: Would it be better to rename the function name to `get_expected_running_containers(self, feature_table)`?

File: src/system-health/health_checker/service_checker.py [hunk 5]
```diff
@@ -24,23 +43,171 @@ class ServiceChecker(HealthChecker):
...
+                        for asic_id in range(num_asics):
+                            containers.add(feature_name + str(asic_id))
+                            container_feature_dict[feature_name + str(asic_id)] = feature_name
+                else:
+                    containers.add(feature_name)
+                    container_feature_dict[feature_name] = feature_name
+
+        return containers, container_feature_dict
+
+    def get_current_running_container_set(self):
```
Comment (yozhao101) [inline src/system-health/health_checker/service_checker.py:84]: Would it be better to rename this function name to `get_current_running_containers(self)`?

File: src/system-health/health_checker/service_checker.py [hunk 6]
```diff
@@ -24,23 +43,171 @@ class ServiceChecker(HealthChecker):
...
+        """Get current running containers, if the running container is not in self.critical_process_dict,
+           try get the critical process list
+
+        Returns:
+            set: A set of running containers
```
Comment (yozhao101) [inline src/system-health/health_checker/service_checker.py:89]: The `set` need be replaced by the variable name `running_containers`?

File: src/system-health/health_checker/service_checker.py [hunk 7]
```diff
@@ -24,23 +43,171 @@ class ServiceChecker(HealthChecker):
...
+        except docker.errors.APIError as err:
+            logger.log_debug("Failed to retrieve the running container list. Error: '{}'".format(err))
+
+        return running_containers
+
+    def get_critical_process_list_from_file(self, container, critical_processes_file):
+        """
+        @summary: Read the critical processes from CRITICAL_PROCESSES_FILE.
+        @return: A list which contain critical processes.
+        """
```
Comment (yozhao101) [inline src/system-health/health_checker/service_checker.py:110]: The fomat of function comments does not align with other functions?

File: src/system-health/health_checker/service_checker.py [hunk 8]
```diff
@@ -24,23 +43,177 @@ class ServiceChecker(HealthChecker):
...
+
+        Returns:
+            critical_process_list: A list of critical process names
+        """
+        critical_process_list = []
+
+        with open(critical_processes_file, 'r') as file:
+            for line in file:
+                # Try to match a line like "program:<process_name>"
+                match = re.match(r"^\s*((.+):(.*))*\s*$", line)
```
Comment (yozhao101) [inline src/system-health/health_checker/service_checker.py:122]: For this regex matching, if the content of variable `line` is `      program   :    lldpd   `, then the `identifier_key` and `identifier_value` will have the preceeding and trailing white spaces, do we need remove them?

File: src/system-health/health_checker/service_checker.py [hunk 9]
```diff
@@ -21,26 +41,174 @@ class ServiceChecker(HealthChecker):
...
-        process and file system.
+        et and analyze the output of $CHECK_CMD, collect status for file system or customize checker if any.
         :param config: Health checker configuration.
         :return:
         """
-        self.reset()
-        output = utils.run_command(ServiceChecker.CHECK_MONIT_SERVICE_CMD).strip()
-        if output != 'active':
+        output = utils.run_command(ServiceChecker.CHECK_MONIT_SERVICE_CMD)
+        if not output or output != 'active':
```
[the remark this answers is not in the document]
Comment (qiluo-msft) [reply src/system-health/health_checker/service_checker.py]: Still applicable.

**Review by qiluo-msft — COMMENTED** [review]
One minor issue remains.

**Comment by Junchao-Mellanox** [comment]
> One minor issue remains.

```python
if not output or output.strip() != 'active':
```

As the output might be None or a string, need check "if not output" here to avoid calling None.strip() method.

**Review by qiluo-msft — APPROVED** [review]

**Review by keboliu — APPROVED** [review]

**Comment by keboliu** [comment]
@yozhao101 would you please approve if no further comments?

File: src/system-health/health_checker/service_checker.py [hunk 10]
```diff
@@ -24,23 +43,177 @@ class ServiceChecker(HealthChecker):
...
+        ctrs = DOCKER_CLIENT.containers
+        try:
+            lst = ctrs.list(filters={"status": "running"})
+
+            for ctr in lst:
+                running_containers.add(ctr.name)
+                if ctr.name not in self.container_critical_processes:
+                    self.fill_critical_process_by_container(ctr.name)
+        except docker.errors.APIError as err:
+            logger.log_debug("Failed to retrieve the running container list. Error: '{}'".format(err))
```
Comment (yozhao101) [inline src/system-health/health_checker/service_checker.py:103]: Do we need change `log_debug(...)` to `log_error(...)`?

File: src/system-health/health_checker/service_checker.py [hunk 11]
```diff
@@ -24,23 +43,177 @@ class ServiceChecker(HealthChecker):
...
+    def fill_critical_process_by_container(self, container):
+        """Get critical process for a given container
+
+        Args:
+            container (str): container name
+        """
+        # Get container volumn folder
+        container_folder = self._get_container_folder(container)
+        if not container_folder:
+            logger.log_debug('Failed to get container folder for {}'.format(container_folder))
```
Comment (yozhao101) [inline src/system-health/health_checker/service_checker.py:145]: `log_debug(...)` to `log_error(...)`?

File: src/system-health/health_checker/service_checker.py [hunk 12]
```diff
@@ -24,23 +43,177 @@ class ServiceChecker(HealthChecker):
...
+            return
+
+        # If container folder does not exist, the container is probably not up, retry it
+        if not os.path.exists(container_folder):
+            logger.log_debug('Container folder does not exist: {}'.format(container_folder))
```
Comment (yozhao101) [inline src/system-health/health_checker/service_checker.py:150]: `log_debug(...)` to `log_error(...)`?

File: src/system-health/health_checker/service_checker.py [hunk 13]
```diff
@@ -24,23 +43,177 @@ class ServiceChecker(HealthChecker):
...
+            return
+
+        # Get critical_processes file path
+        critical_processes_file = os.path.join(container_folder, ServiceChecker.CRITICAL_PROCESSES_PATH)
+        if not os.path.isfile(critical_processes_file):
+            # Critical process file does not exist, the container has no critical processes.
+            # This is fine, don't retry.
+            logger.log_debug('Failed to get critical process file for {}, {} does not exist'.format(container, critical_processes_file))
```
Comment (yozhao101) [inline src/system-health/health_checker/service_checker.py:158]: `log_debug(...)` to `log_error(...)`?

File: src/system-health/health_checker/service_checker.py [hunk 14]
```diff
@@ -70,3 +243,100 @@ def check(self, config):
...
+
+        for bad_container in self.bad_containers:
+            self.set_object_not_ok('Service', bad_container, 'Syntax of critical_processes file is incorrect')
+
+    def check(self, config):
+        """
+        Check critical system service status.
+        :param config: Health checker configuration.
+        :return:
+        """
```
Comment (yozhao101) [inline src/system-health/health_checker/service_checker.py:285]: The function comment did not align with other functions.

File: src/system-health/health_checker/service_checker.py [hunk 15]
```diff
@@ -70,3 +243,100 @@ def check(self, config):
...
+            items = line.split()
+            if len(items) < 2:
+                continue
+            data[items[0].strip()] = items[1].strip()
+        return data
+
+    def check_process_existence(self, container_name, critical_process_list, config, feature_table):
+        """
+        @summary: Check whether the process in the specified container is running or not.
+        """
```
Comment (yozhao101) [inline src/system-health/health_checker/service_checker.py:313]: The function comment did not align with other functions.

File: src/system-health/health_checker/service_checker.py (hunk 13 above)
Comment (Junchao-Mellanox) [reply src/system-health/health_checker/service_checker.py:158]: Debug log is fine here. Some container might not have critical processes file, this is allowed.

**Review by keboliu — DISMISSED** [review]

**Review by qiluo-msft — DISMISSED** [review]

**Review by yozhao101 — DISMISSED** [review]

**Comment by Junchao-Mellanox** [comment]
The checker failed due to yang model unit test failure. I will try rebase it.

**Comment by qiluo-msft** [comment]
Could you rebase to latest master?

**Comment by Junchao-Mellanox** [comment]
No clean cherry-pick for 202106/202012, need separate PRs.

**Review by qiluo-msft — APPROVED** [review]

**Comment by yozhao101** [comment]
> No clean cherry-pick for 202106/202012, need separate PRs.

Can you submit separate PRs against 202012/202106 please?

**Comment by Junchao-Mellanox** [comment]
Hi @yozhao101, I suppose they are already in 202012 and 202106. Please check https://github.com/Azure/sonic-buildimage/pull/9366 and https://github.com/Azure/sonic-buildimage/pull/9367

**Pull request merged**